### ISSUE TITLE
CI always reports escape analysis results

### ISSUE THREAD

**Issue by yoheiueda** [confidential-containers/cloud-api-adaptor#604]
Currently, CI always report results of escape analysis for `main.go` as shown in the attached image.

These escaping variables do not mean bugs or memory leaks. These information are just hints for performance optimizations.  In my opinion, reporting these results in CI seem confusing, since they are shown as failures. I think it is safe for us to disable escape checks in CI.

<img width="820" alt="warnings" src="https://user-images.githubusercontent.com/9316614/218036100-37c72fd0-7bbb-4cfa-ad19-a139a03aa9a4.png">

**Comment by yoheiueda** [comment]
@SamYuan1990 do you have any comments?

**Comment by SamYuan1990** [comment]
It depends, some of memory leak is due to golang language itself. But it may also cause by bug or code in bad implementation.
Memory leak will cost more GC happens.
so I am agree it is not as failures, and as hints for performance optimizations.
here is a sample: https://dave.cheney.net/2014/06/07/five-things-that-make-go-fast
@yoheiueda could you please try to test with escape analysis sample function in a pr of this repo to see if we can cover escape analysis test by other CI processing or not?
if not, I think it is valuable. but I would like to follow team's final decision.

**Comment by yoheiueda** [comment]
@SamYuan1990 Thank you for the reply.

I prepared a draft PR for discussion. #609 

One thing I noticed is that the current test only reports escape checks in `main.go` only. I don't know why the other files are not checked.  Performance optimizations in `main.go` does not improve the overall performance in our case.

Another point is that the currently test results are filtered with `grep`, but the output of escapes check is printed in stderr, so I think we need to fix it like this https://github.com/confidential-containers/cloud-api-adaptor/pull/609/commits/0c75842d93b5b215ff9f3b78c97e40825b551979.

**Comment by SamYuan1990** [comment]
> @SamYuan1990 Thank you for the reply.
> 
> I prepared a draft PR for discussion. #609
> 
> One thing I noticed is that the current test only reports escape checks in `main.go`. I don't know why the other files are not checked. Performance optimizations in `main.go` will not improve the overall performance in our case.
> 
> Another point is that the currently test results are filtered with `grep`, but the output of escapes check is printed in stderr, so I think we need to fix it like this [0c75842](https://github.com/confidential-containers/cloud-api-adaptor/commit/0c75842d93b5b215ff9f3b78c97e40825b551979).

here is a good sample https://github.com/confidential-containers/cloud-api-adaptor/pull/609/files#r1104452743

**Comment by yoheiueda** [comment]
@SamYuan1990 

I still see escapes check results in `main.go` only in the **Files changed** page as described at https://github.com/confidential-containers/cloud-api-adaptor/pull/610#issuecomment-1429056174

There are more than 2,000 errors per cloud provider, and I think showing them in **Files changed** seems too excessive.

**Comment by yoheiueda** [comment]
In my opinion, CAA is not a performance sensitive program for now, and fixing the potential escapes may lead to [premature optimizations](https://stackify.com/premature-optimization-evil/). It may degrade code readability and maintainability. 

Escaped variables will not always degrades GC performance. In a typical case, GC performance becomes critical when long-lived objects are bloated. Short-lived objects, and a small amount of long-lived objects are not problems for memory management.

When the performance of CAA matters, then we will profile CAA, and identify hot functions and memory allocation sites. As you know, software usually has the 80-20 rule, so optimizing the hot 20% of a program will lead to a significant performance improvements.

For the resaons above, I don't think the 2,000+ errors in CI results is not helpful to improve the code quality. @SamYuan1990 what do you think?

**Comment by SamYuan1990** [comment]
> In my opinion, CAA is not a performance sensitive program for now, and fixing the potential escapes may lead to [premature optimizations](https://stackify.com/premature-optimization-evil/). It may degrade code readability and maintainability.
> 
> Escaped variables will not always degrades GC performance. In a typical case, GC performance becomes critical when long-lived objects are bloated. Short-lived objects, and a small amount of long-lived objects are not problems for memory management.
> 
> When the performance of CAA matters, then we will profile CAA, and identify hot functions and memory allocation sites. As you know, software usually has the 80-20 rule, so optimizing the hot 20% of a program will lead to a significant performance improvements.
> 
> For the resaons above, I don't think the 2,000+ errors in CI results is helpful to improve the code quality. @SamYuan1990 what do you think?

well, I don't want to repeat myself too many times. but
- I will follow team's decision.
- No matter what decision is, 20-80 or ... is a balance between the advantages and disadvantages.
- As initiator, I suppose my response here is to explain what it is.(The facts, as I don't want to have any discussion with any against style, or we force to do sth)

Here are the facts:
- Some memory leak is due to golang implementation itself, for example big number packages or when we use []string as parameter input of a function. Hence for each line, we need to check it case by case, as some times it is not an error.
- Memory leak test can help us find out some implementations which are not good, as it leak to memory. I am not sure if golang can been trusted as a memory safe coding language. Here I just want to point out a question, what's memory leak at security point of view?
- CAA is not performance matters appliance, just when it is not used to CURD with pod in high frequency? hence even if pod auto scale is an edge case right? If all normal cases are good, we can trade off on edge cases.

As a summary, @yoheiueda 
May I learn from this ticket, I suppose we don't have to close it in this week or this month before we figure out:
- do we want to benefits from escaped checking?
- is golang trusted as memory safe language?
- CAA is performance matters or not? If there any case performance matter, what is that?
- For this project, what's the trade off?

I suppose 2000 lines or 3000 lines... the number is not the key point.

**Comment by huoqifeng** [comment]
How about keep the `escape` check in `Makefile` and disable it in CI git actions? So that developers can run/fix it by themselves. Then we can enable the `escape` check in future in CI? I talked with @SamYuan1990 , he agree with this. @yoheiueda thoughts?

**Comment by yoheiueda** [comment]
> How about keep the `escape` check in `Makefile` and disable it in CI git actions? So that developers can run/fix it by themselves. Then we can enable the `escape` check in future in CI? I talked with @SamYuan1990 , he agree with this.

@huoqifeng I agree. My PR was prepared in the direction. https://github.com/confidential-containers/cloud-api-adaptor/pull/609

I'll update the PR to show results for all files with the Makefile check.

**Comment by bpradipt** [comment]
> How about keep the `escape` check in `Makefile` and disable it in CI git actions? So that developers can run/fix it by themselves. Then we can enable the `escape` check in future in CI? I talked with @SamYuan1990 , he agree with this. @yoheiueda thoughts?

I think this approach makes sense for the time being. 

We do need to start focussing on the performance aspects especially w.r.t pod VM startup time as that's the one which is immediately visible to the end-user.
For reference linking the related issue here -  https://github.com/confidential-containers/cloud-api-adaptor/issues/343
